### PR TITLE
Egress SNAT for egress traffic from workload based on an annotation on the pod

### PR DESCRIPTION
**Pull request by sriramy**
## Description

As I mentioned in the issue 3275, this proposal provides a consistent egress IP for outbound connections from the pod without the need for an edge router.

A proposal to apply SNAT on egress traffic from workload towards external network (IP address that doesn't fall in any of the calico IP pools). Some issues that this addresses:
https://github.com/projectcalico/calico/issues/3275
https://github.com/projectcalico/calico/issues/5549

I tested this with Felix FV with the following tests:
- ETCD datastore with linux dataplane
- ETCD datastore with eBPF dataplane
- Kubernetes datastore with linux dataplane (Fails, and not sure why)
- Kubernetes datastore with eBPF dataplane

## Related issues/PRs

https://github.com/projectcalico/calico/issues/3275
https://github.com/projectcalico/calico/issues/5549

## Todos

- [ ] Tests
- [ ] Documentation
- [ ] Release note

## Release Note

```release-note
TBD
```

## Reminder for the reviewer

Make sure that this PR has the correct labels and milestone set.

Every PR needs one `docs-*` label.

- `docs-pr-required`: This change requires a change to the documentation that has not been completed yet.
- `docs-completed`: This change has all necessary documentation completed.
- `docs-not-required`: This change has no user-facing impact and requires no docs.

Every PR needs one `release-note-*` label.

- `release-note-required`: This PR has user-facing changes. Most PRs should have this label.
- `release-note-not-required`: This PR has no user-facing changes.

Other optional labels:

- `cherry-pick-candidate`: This PR should be cherry-picked to an earlier release. For bug fixes only.
- `needs-operator-pr`: This PR is related to install and requires a corresponding change to the operator.

### CODE REVIEW

File: libcalico-go/lib/apis/v3/workloadendpoint.go [hunk 1]
```diff
@@ -65,6 +65,10 @@ type WorkloadEndpointSpec struct {
 	// to connect one of its own external IPs. Each internal IP must be associated with the same
 	// endpoint via the configured IPNetworks.
 	IPNATs []IPNAT `json:"ipNATs,omitempty" validate:"omitempty,dive"`
+	// Egress SNAT is a 1:1 NAT mapping to apply for oubtound connections from the endpoint.
```
Comment (neiljerram) [inline libcalico-go/lib/apis/v3/workloadendpoint.go:68]: sp, should be "outbound"

File: felix/dataplane/linux/masq_mgr.go [hunk 2]
```diff
@@ -1,4 +1,4 @@
-// Copyright (c) 2016-2021 Tigera, Inc. All rights reserved.
+// Copyright (c) 2016-2021 Tigera, Inc. All rights reserved.Address
```
Comment (neiljerram) [inline felix/dataplane/linux/masq_mgr.go:1]: accidental paste here

File: felix/dataplane/linux/egress_snat_mgr.go [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+// "masquerade" feature.  The feature adds a boolean flag to each IPAM pool, which controls how
+// outgoing traffic to non-Calico destinations is handled.  If the "masquerade" flag is set,
+// outgoing traffic is source-NATted to appear to come from the host's IP address.
+//
+// The egressSNATManager maintains two CIDR IP sets: one contains the CIDRs for all Calico
+// IPAM pools, the other contains only the NAT-enabled pools.
+//
+// When NAT-enabled pools are present, the egressSNATManager inserts the iptables masquerade rule
+// to trigger NAT of outgoing packets from NAT-enabled pools.  Traffic to any Calico-owned
+// pool is excluded.
```
Comment (neiljerram) [inline felix/dataplane/linux/egress_snat_mgr.go:37]: Comment mostly wrong IIUC, needs rewriting from scratch.

File: felix/Makefile [hunk 4]
```diff
@@ -275,7 +275,7 @@ ut combined.coverprofile: $(SRC_FILES) $(LIBBPF_A) build-bpf
 fv/fv.test: $(LIBBPF_A) $(SRC_FILES) $(FV_SRC_FILES)
 	# We pre-build the FV test binaries so that we can run them
 	# outside a container and allow them to interact with docker.
-	$(DOCKER_GO_BUILD_CGO) go test $(BUILD_FLAGS) ./$(shell dirname $@) -c --tags fvtests -o $@
+	$(DOCKER_GO_BUILD_CGO) go test $(BUILD_FLAGS) ./$(shell dirname $@) -c --tags nattests -o $@
```
Comment (neiljerram) [inline felix/Makefile:278]: Do we still want fvtests as well?

Comment (sriramy) [reply felix/Makefile]: This was supposed to be reverted, done now.

File: felix/dataplane/linux/egress_snat_mgr.go (hunk 3 above)
Comment (sriramy) [reply felix/dataplane/linux/egress_snat_mgr.go]: Done.

File: felix/dataplane/linux/masq_mgr.go (hunk 2 above)
Comment (sriramy) [reply felix/dataplane/linux/masq_mgr.go]: Removed.

File: libcalico-go/lib/apis/v3/workloadendpoint.go (hunk 1 above)
Comment (sriramy) [reply libcalico-go/lib/apis/v3/workloadendpoint.go]: Done.

**Comment by sriramy** [comment]
Just a note that this solution has outstanding issues to be fixed, more here: [3275](https://github.com/projectcalico/calico/issues/3275#issuecomment-1288922328)

**Comment by caseydavenport** [comment]
@sriramy thanks for the PR! 

We're discussing this one a bit within the maintainers team at the moment. We need to work through a few scenarios to make sure this solution is comprehensive enough and is something we are able to merge. We'll report back here with the outcome of that discussion.

I also saw your comment here https://github.com/projectcalico/calico/issues/3275#issuecomment-1288922328 - we will think about that. The multi-node scenario is one of the key limitations that we need to understand and that we were discussing earlier as well.

File: felix/rules/nat.go [hunk 5]
```diff
@@ -151,6 +151,62 @@ func (r *DefaultRuleRenderer) SNATsToIptablesChains(snats map[string]string) []*
...
+		extIp := snats[intIp]
+		rules = append(rules, iptables.Rule{
+			Match:  iptables.Match().NotDestIPSet(allIPsSetName).SourceNet(intIp),
+			Action: iptables.SNATAction{ToAddr: extIp},
+		})
+	}
+	return []*iptables.Chain{{
+		Name:  ChainNATEgress,
+		Rules: rules,
+	}}
```
Comment (tomastigera) [inline felix/rules/nat.go:207]: This seems to be duplicated by egressSNATsToIptablesChainsBPF. What if we just have the if bpf enabled here?

Comment (sriramy) [reply felix/rules/nat.go]: Done. I had copied what was done in MakeNatOutgoingRule, but I agree it is better to check BPFEnabled in the same function.

**Comment by sriramy** [comment]
> @sriramy thanks for the PR!
> 
> We're discussing this one a bit within the maintainers team at the moment. We need to work through a few scenarios to make sure this solution is comprehensive enough and is something we are able to merge. We'll report back here with the outcome of that discussion.
> 
> I also saw your comment here [#3275 (comment)](https://github.com/projectcalico/calico/issues/3275#issuecomment-1288922328) - we will think about that. The multi-node scenario is one of the key limitations that we need to understand and that we were discussing earlier as well.

Hi, wondering if there was some conclusion on this topic? :)

**Comment by caseydavenport** [comment]
@sriramy sorry for the long delay, we've been a bit slow over the holidays. I don't think there's been much traction on this one I'm afraid.

**Comment by mgleung** [comment]
Hey @sriramy, sorry for the delay still, I think we are still missing some of the discussion needed for this so I'll re-ping the team to get the conversation rolling.

**Comment by sriramy** [comment]
No worries, we are still interested in the feature though. I would like to contribute if some way forward is suggested. :) Do you think something like Cilium Egress Gateway[1] or Antrea Egress[2] can be a better idea than what I have suggested above? 
[1]: https://docs.cilium.io/en/v1.10/gettingstarted/egress-gateway/
[2]: https://antrea.io/docs/v1.3.0/docs/egress/

**Comment by xagent003** [comment]
I would like to see this scenario as well. We use Calico BGP to advertise an external IP for a service. It works great for ingress, the Pod is reachable by the external IP. For  egress it gets SNAT'd using the nodes IP. Would like for it to go out with the same external IP used for BGP advertisements and ingress.

For our usecase these are actually kubevirt VMs so there is only 1 Pod endpoint

There are some egress gateway solutions(Calico Enterprise has one) but thats not what I was looking for. That means all pods go to one egress vrouter and get NATd out using the same IP.

In our case each Pod (VM) has its own external IP that is externally routaBle, It may not always be the case theres an edge gateway for the nodes (so no Internet access, at least via the Calico interface). Our nodes only have Internet via a separate mgmt NIC. This is why node SNAT may not work.

As an example w/ openstack neutron, when a VM has a floating IP, it ingresses and egresses with the floating IP. The router does NAT both directions. It isnt: `"ingress on floating, but egress using node or some other  egress gateway"` Only VMs without a floating IP share a common snat egress gateway

In this K8s case the "floating IP" is our external IP on the Service(or external Loadbalancer IP)

**Comment by mgleung** [comment]
Hey @sriramy, sorry for the delay again, I'll definitely keep you updated as we progress on this. The members of the team that would have the best insight into this have been focused on some other feature work so I'll try to get their attention on this. Thanks for your patience and we're looking forward to working with you on this.

@xagent003 could I ask you to clarify this portion of your use case?
> In our case each Pod (VM) has its own external IP that is externally routaBle, It may not always be the case theres an external gateway for the nodes (so no Internet access, at least via the Calico interface). Our nodes only have Internet via a separate mgmt NIC. This is why node SNAT may not work for VMs/Pods needing Internet.
My understanding is that if everything has to go through the separate NIC on your nodes, traffic would leave with that IP which seems pretty similar to the case where we're using the node's IP. Is the issue that those IPs differ and you're looking to choose one over the other?

**Comment by xagent003** [comment]
> My understanding is that if everything has to go through the separate NIC on your nodes, traffic would leave with that IP which seems pretty similar to the case where we're using the node's IP. Is the issue that those IPs differ and you're looking to choose one over the other?

So, this is a multi-tenant multi-cluster cloud so all the nodes are locked down in different VLANs and VRFs at the TORs to provide isolation between different clusters. I am told right now there's an edge gateway device per, but, they want to get rid of it as it won't scale. On the other hand, another "mgmt interface" does exist, but its connected via an entirely different means, no K8s traffic should use it. I'm not a data center expert, so hope that made sense... But for now ignore it, imagine it doesn't exist :)

Now, suppose bond1 has an IP of 10.240.0.34. This is the interface we use for Calico. Traffic outbound gets SNAT'd using 10.24.0.34. Except there is no public gateway for this (it does not route to the Internet). Now we do have public IP space. We set up Calico BGP to peer with the TORs then assign an explicit public IP to the service like such:

```
apiVersion: v1
kind: Service
metadata:
  name: netdebug4
  annotations:
    metallb.universe.tf/loadBalancerIPs: 1.2.3.4
spec:
  selector:
    ngpc.pf9.io/vmService: vmcalicobridge2
  ports:
    - protocol: TCP
      port: 22
  externalTrafficPolicy: Local
  type: LoadBalancer
  #externalIPs:
    #- 1.2.3.4
 ```

The BGP works great, using either the externalIP on the Service, or using MetalLB (we still use Calico for the BGP, metalLB is only so we can set type: LoadBalancer and externalTrafficPolicy: Local). Either way above Service gets a public IP of 1.2.3.4. Inbound traffic is routed down to bond1 for the public IP, 1.2.3.4 in above example. And then kube-proxy NATs it to the Pod (Kubevirt VM).

On egress I wanted to also use 1.2.3.4 as source IP. Not any node IP.

The containers CIDR is not routable. The Services CIDR is not routable. The node IPs potentially not either. Which means egress(public, to the Internet) won't work unless it goes out using the public IP

**Comment by mgleung** [comment]
hey @sriramy and @xagent003 , I touched base with the team and while there was a lot of discussion, I believe that one of the major hold ups is that we're not exactly sure how the changes in this PR ensure that reply traffic gets routed back to the pod. For floating IPs, Felix programs a local device route for each floating IP and BIRD propagates them in the same way it propagates pod IPs. To help us understand how this is supposed to work for these changes, maybe we need to back up and ask:

1. What is the pattern of SNAT IPs that are going to be assigned to pods when using this feature?
2. How do we arrange that traffic to one of those SNAT IPs (coming from outside the cluster) will be routed to the correct node?

**Comment by sriramy** [comment]
> hey @sriramy and @xagent003 , I touched base with the team and while there was a lot of discussion, I believe that one of the major hold ups is that we're not exactly sure how the changes in this PR ensure that reply traffic gets routed back to the pod. For floating IPs, Felix programs a local device route for each floating IP and BIRD propagates them in the same way it propagates pod IPs. To help us understand how this is supposed to work for these changes, maybe we need to back up and ask:
> 
> 1. What is the pattern of SNAT IPs that are going to be assigned to pods when using this feature?
> 2. How do we arrange that traffic to one of those SNAT IPs (coming from outside the cluster) will be routed to the correct node?

Hi @mgleung, This PR was just an RFC patch really.
For a full solution, I was hoping to create a IPPool object for IP addreses that can be used for SNAT so the routes are exported via BGP. We could add a new allowedUse type called "SNAT" or something similar.

**Comment by xagent003** [comment]
Hi matt, not sure I understand your questions...

I dont understand #1 what you mean by pattern? As I mentioned, we want it to be SNAT'd using the same externalIP assigned to the service. So packets will ingress and egress using the same IP. We take a public IPv4 address and add it to a Service as the externalIP

For #2, doesn't BGP take care of it? We advertise the external IP associated with a Service. The packet ingress's to that node. thereafter, kube-proxy "loadbalances" it to the Pod. Which in our case is a singular Kubevirt VM. Very likely on the same node because we want to use externalTrafficPolicy: local.

On egress, it would take the exact same path outbound... except it gets SNAT'd using the Service's external IP rather than the node IP on the NIC.

**Comment by xagent003** [comment]
@mgleung @sriramy OVN-Kubernetes seems ot have something like this: https://github.com/ovn-org/ovn-kubernetes/blob/master/docs/egress-service.md

`The Egress Service feature enables the egress traffic of pods backing a LoadBalancer service to exit the cluster using its ingress IP. This is useful for external systems that communicate with applications running on the Kubernetes cluster through a LoadBalancer service and expect that the source IP of egress traffic originating from the pods backing the service is identical to the destination IP they use to reach them - i.e the LoadBalancer's ingress IP.`

**Comment by caseydavenport** [comment]
Thank you everyone for the interest and discussion above, and apologies this hasn't come sooner. I think for now I am going to have to close this PR. There are a few reasons at play here:

- Our enterprise product has committed to a different approach to solving this problem via the use of egress gateways. 
- We aren't fully sure this approach is a complete solution. Specifically the concerns around multi-node clusters and reliance on host IP addresses for SNAT require additional work to make viable.
- This type of feature is - for now - out of scope for the Calico opens-source project, so our ability to invest time in fully building out this feature set is too limited.

Again, thanks all for your input and contributions above.

**Pull request closed without merging**